man/httpdc.1
.Dd May 29, 1996
.Dt HTTPDC 1
.Os xs-httpd/3.5
.Sh NAME
.Nm httpdc
.Nd Control program for the xs-httpd webserver
.Sh SYNOPSIS
.Nm httpdc
.Op Fl d Ar rootdir
.Op Fl p Ar pidfile
.Op Fl v
.Op Ar command
.Sh DESCRIPTION
.Ql httpdc
is a program that lets you control the behaviour of the
xs\-httpd daemon.
With this program, you can retrieve its status, terminate
the webserver, restart it, etc.
.Pp
Without any arguments, the program will start in interactive
mode showing a prompt in which you can issue multiple
commands.
If a command is given on the command line, the program will
execute that command and terminate immediately when it is
finished.
.Pp
The following control commands are available,
.Bl -tag -width Ds -compact
.It Ic ?
.It Ic help
Gives a list of command and their meanings.
.Pp
.It Ic status
Display the webserver's status. It checks whether the main
daemon is running and if the process group is still active.
It also displays the pid of the main process and the
command that was used to start the webserver.
.Pp
.It Ic stop
Shutdown the webserver. It will only stop the main daemon:
that daemon will shut down its child processes before exiting.
.Pp
.It Ic kill
Forcefully kills the webserver and all child processes.
This should only be used if a normal
.Ic stop
command isn't working or doesn't respond fast enough.
.Pp
.It Ic restart
Shuts the webserver down, waits until it is completely dead
and then restarts it using the command line arguments that
were used last time it was started. If the daemon was not
running, it will print a warning and start it anyway.
.Pp
.It Ic reload
This makes the webserver reload its global configuration files
.Pq Pa httpd.conf , mime.types No and Pa compress.methods .
Note that this will not redo the complete configuration; in
particular this will not create or terminate any new sockets
that have been added to or removed from the
.Xr httpd.conf 5
- nor will it change the owner (uid/gid) of running
processes. In general it is better to use
.Ic restart
to ensure that the complete new configuration has been
activated.
.Pp
.It Ic quit
.It Ic exit
Quits the control program. You may also press
.Li CTRL-D
(or any other key that means
.Ql end of input )
to quit the program.
.El
.Sh SEE ALSO
.Xr httpd 1 ,
.Xr httpd.conf 5
.Pp
The project homepage:
.Pa http://www.xs\-httpd.org/


ChangeLog
This file contains an overview of version differences. The list
is not complete. Refer to cvs comments for more information.

Changes since 3.7 beta 9
- Add SSLCipherList option - see ciphers(1)
- Missing User-Agent or Referer will be logged as "-" rather than ""
- Authenticated usernames will be recorded in access_log

Changes since 3.7 beta 8
- New ServerInfo option to set version string (full, branch, name)
- Pass along all HTTP headers when a forwarding proxy request

Changes since 3.7 beta 7
- Rename hier.layout to dirs.conf and shuffle configurable dirs

Changes since 3.7 beta 6
- Fix IconDir/ExecDir usage for vhosts
- Fix handling of unprivileged uid
- Fix mod_python for versions other than 2.5
- Fix mod_gzip dependency
- Fix mod_ruby initialization (enable component loading)
- Easier install directory configuration via hier.layout
- Abort configure when a selected package cannot be used

Changes since 3.7 beta 5
- Execute in .xsconf now works if script.methods is emtpy
- UseLocalScript now defaults to on (and is checked properly)
- Fix on-the-fly (de)compression: now works with large files

Changes since 3.7 beta 4
- Add UseGzipCompression to mod_gzip for on-the-fly compression
- Build mod_bzip2 by default

Changes since 3.7 beta 3
- Fix bug that caused internal: commands in .xsscripts to be ignored
- Add mod_gzip, mod_compress, mod_bzip to replace external binaries
- Add Content-MD5 trailers for CGI scripts

Changes since 3.7 beta 2
- Fix authentication module logic: all modules need to grant access

Changes since 3.7 beta 1
- Modules are now implemented as shared libraries
- Add Modules option to set which modules get loaded

Changes since 3.6 beta 24
- Fix LDAP authentication
- Introduce seperated modules (mod_ldap, perl, python, ruby)
- Make OpenSSL mandatory (removes 40 #ifdefs, 2 libcompat files)

Changes since 3.6 beta 23
- Clean up 'forward' syntax in .redir to match 'redir' command
- Include a replacement crypt() if libc, libcrypt and libssl fail
- Add NoPrivs option to .xsconf (and deprecate .xsuid)
- Enable Smashing Stack Protector (SSP) by default
- Fix parsing of empty lines in mime.types

Changes since 3.6 beta 22
- Fix possible build conflict caused by libmd and openssl
- Fix parsing of mime.types files

Changes since 3.6 beta 21
- Fix memory mismanagement in digest authentication verification
- Replace fgets(), fgetln() by friendlier fparseln() routine
  Include replacement functions for fgetln() and fparseln()
- Fix crash in https timeout handler
- Portability fixes (b20 doesn't work on many platforms)
- Tweak TCP options (especially for Linux systems)
- Add xspasswd -r option to remove accounts from password file

Changes since 3.6 beta 20
- Remove arbitrary string length limits for many configuration files,
  HTTP header messages and SSL certificate fields
- Add support for SSLVhosts (using TLS extension: Server Name Indication)
- Fix NULL-pointer dereference and string termination in HTML encoding
- Add UseSendfile option in httpd.conf to disable sendfile(2)

Changes since 3.6 beta 19
- Fix memory mismanagement in digest authentication verification

Changes since 3.6 beta 18
- Enable experimental ruby interpreter (internal:ruby)

Changes since 3.6 beta 17
- Fix parsing problem in run-cgi SSI (and maybe other places)

Changes since 3.6 beta 16
- Add index.xhtml and index.xml to default IndexFiles
- Improve url-encoding routine

Changes since 3.6 beta 15
- Fix escaping of user input used in HTML error messages;
  possible XSS abuse reported by Serge van den Boom
- Fix checks for pcre used by libxs

Changes since 3.6 beta 14
- Fix potential crash in server-side include parsing

Changes since 3.6 beta 13
- Always terminate SSL connection when reading from socket fails
- Add conditional ifenv option to .redir files

Changes since 3.6 beta 12
- Don't install empty docdir
- Shuffle internal structures (reduce global vars, getenv calls)
- Reject non-HTTP protocol specifiers

Changes since 3.6 beta 11
- Drop persistent connections after client read/write errors

Changes since 3.6 beta 10
- Add RedirFile configuration option for light-weight vhosts
- Sort qvalues in accept-* arrays
- Fake qvalues for */* and type/* when qvalues are not used

Changes since 3.6 beta 9
- Use original URL in rewriting rules before trying index.html
- Add PhIconDir option, analogous to PhExecDir

Changes since 3.6 beta 8
- Close unused CGI pipes (could block execution), by jilles

Changes since 3.6 beta 7
- Extend characters available for alternative setproctitle()
- Lock pidfile and don't start if somebody holds a lock
- Allow multiple arguments to Restrict and AuthFile
- Allow method selector in .xsconf blocks (e.g. only match POST)

Changes since 3.6 beta 6
- Improve configuration checks
- Fix include order and typechecks

Changes since 3.6 beta 5
- Enable FastCGI support again (experimental)
- Clean Perl and Python detection and initialisation (experimental)

Changes since 3.6 beta 4
- Shuffle configuration files
- Fix build outside source directory (wrt config.h)
- Fix sendfile on Linux
- Fix infinate read when content-length exceeds post data
- Portability fixes in function/library checking

Changes since 3.6 beta 3
- Remove outputbuffer and set TCP_NOPUSH (where available)
- Add 'passexist' redirection command

Changes since 3.6 beta 2
- Fix server-side include variable cleanup
- Use sendfile (where appropriate)

Changes since 3.6 beta 1
- Add http status code back to $ERROR_READABLE

Changes since 3.5 beta 31
- No functional changes

Changes since 3.5 beta 30
- Fix read errors with long POST content

Changes since 3.5 beta 29
- Make CGI header size dynamic (unlimited)
- Add default PathInfoScripts (imagemap, xschpass)
- Update manpages

Changes since 3.5 beta 28
- Fix basic authentication check
- Fix ldap authentication check

Changes since 3.5 beta 27
- Fix reading of stderr data from CGI scripts (could block CGI)
- Minor code cleanups

Changes since 3.5 beta 26

src/httpdc.c
/* Copyright (C) 1995, 1996 by Sven Berkvens (sven@stack.nl) */
/* Copyright (C) 1998-2008 by Johan van Selst (johans@stack.nl) */

#include	"config.h"

#include	<sys/types.h>
#include	<signal.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<stdlib.h>
#ifdef		HAVE_ERR_H
#include	<err.h>
#endif		/* HAVE_ERR_H */
#include	<errno.h>
#include	<string.h>
#include	<ctype.h>

#include	"malloc.h"
#include	"path.h"

typedef	struct
{
	const	char	*command;
	void		(*func) (const char *);
	const	char	*help;
} command;

static	pid_t	httpdpid;	/* a value of 0 denotes no running httpd */
static	char	startparams[BUFSIZ];
char		rootdir[XS_PATH_MAX];

static	void	cmd_help	(const char *);
static	void	cmd_status	(const char *);
static	void	cmd_kill	(const char *);
static	void	cmd_stop	(const char *);
static	void	cmd_start	(const char *);
static	void	cmd_reload	(const char *);
static	void	cmd_restart	(const char *);
static	void	cmd_version	(const char *);
static	void	control		(const char *);
static	void	loadpidfile	(const char *);
static	void	getpidfilename	(char **);

static	command	commands[]=
{
	{ "?",		cmd_help,	"Display this help text"	},
	{ "help",	cmd_help,	"Display this help text"	},
	{ "status",	cmd_status,	"Display httpd status"		},
	{ "start",	cmd_restart,	"Start httpd"			},
	{ "restart",	cmd_restart,	"Restart httpd"			},
	{ "reload",	cmd_reload,	"Reload all httpd databases"	},
	{ "stop",	cmd_stop,	"Terminate the httpd"		},
	{ "kill",	cmd_kill,	"Forcefully terminate the httpd"},
	{ "version",	cmd_version,	"Show httpdc version string"	},
	{ "quit",	NULL,		"Quit the control program"	},
	{ "exit",	NULL,		"Quit the control program"	},
	{ NULL,		NULL,		NULL				},
};

static	void
cmd_help(const char *args)
{
	command		*search;

	for (search = commands; search->command; search++)
		printf("%s\t\t%s\n", search->command, search->help);
	(void)args;
}

static	void
cmd_status(const char *args)
{
	if (!httpdpid)
		return;

	if (kill(httpdpid, 0))
	{
		if (errno == ESRCH)
			warnx("Main HTTPD does not seem to be running");
		else
			warn("kill()");
	}
	else
		printf("Main HTTPD seems to be running\n");

	if (killpg(httpdpid, 0))
	{
		if (errno == ESRCH)
			warnx("HTTPD process group does not seem to be running\n");
		else
			warn("killpg()");
	}
	else
		printf("HTTPD process group seems to be running\n");

	printf("Main HTTPD PID: %ld\n", (long)httpdpid);
	printf("Last used command line: %s\n", startparams);
	(void)args;
}

static	void
cmd_stop(const char *args)
{
	if (!httpdpid)
		return;

	if (kill(httpdpid, SIGTERM))
		warn("kill()");
	else
		printf("Main HTTPD terminated, children will die too.\n");
	(void)args;
}

static	void
cmd_kill(const char *args)
{
	int	timeout;

	if (!httpdpid)
		return;

	timeout = 600;
	printf("Killing HTTPD processes... ");
	while (!killpg(httpdpid, SIGKILL) && (timeout > 0))
	{
		printf("%c\b", "/-\\|"[timeout & 3]);
		fflush(stdout);
		sleep(1);
		timeout--;
	}
	if (!killpg(httpdpid, 0))
		warnx("The children would not die within %d seconds!", timeout);
	else
		printf("All children have been killed.\n");
	(void)args;
}

static	void
cmd_restart(const char *args)
{
	int		timeout = 600;

	if (!httpdpid ||
		(kill(httpdpid, 0) && errno == ESRCH))
	{
		/* server not running: goto start */
		cmd_start(args);
		return;
	}

	if (kill(httpdpid, SIGTERM))
		warn("kill()");
	printf("Main HTTPD killed, children will die automatically.\n");
	printf("Waiting for children to die... ");

	while (!killpg(httpdpid, 0) && (timeout > 0))
	{
		printf("%c\b", (char)*("/-\\|" + (timeout & 3)));
		fflush(stdout);
		sleep(1);
		timeout--;
	}

	if (!killpg(httpdpid, 0))
	{
		warnx("The children would not die within %d seconds!\n",
			timeout);
		return;
	}
	printf("Children are dead!\n");
	cmd_start(args);
}

static	void
cmd_start(const char *args)
{
	if (!httpdpid)
		printf("Starting with default options... ");
	else
		printf("Restarting httpd... ");

	fflush(stdout);
	system(startparams);
	printf("Done!\n");
	printf("Executed: %s\n", startparams);
	(void)args;
}

static	void
cmd_reload(const char *args)
{
	if (!httpdpid)
		return;

	if (kill(httpdpid, SIGHUP))
		warn("kill()");
	else
		printf("Databases reloaded...\n"
			"Run 'httpdc restart' to ensure that "
			"all configuration changes take effect.\n");

	(void)args;
}

static	void
cmd_version(const char *args)
{
	char	*pidfile;

	getpidfilename(&pidfile);
	printf("%s\n", SERVER_IDENT);
	printf("Pid file: %s\n", pidfile);
	(void)args;
}

static	void
control(const char *args)
{
	char		buffer[BUFSIZ], *space;
	command		*search;

	strlcpy(buffer, args, BUFSIZ);
	space = buffer;
	while (*space && (*space != 9) && (*space != ' '))
		space++;
	if (*space)
	{
		*(space++) = 0;
		while ((*space == 9) || (*space == ' '))
			space++;
	}
	search = commands;
	while (search->command)
	{
		if (!strcasecmp(search->command, buffer))
		{
			if (search->func)
				search->func(space);
			else
				exit(0);
			return;
		}
		search++;
	}
	warnx("Command `%s' not found\n", buffer);
}

static	void
getpidfilename(char **pidfilename)
{
	char		*p, buffer[BUFSIZ], config_path[XS_PATH_MAX];
	FILE		*conffile;

	if (!pidfilename)
		return;

	*pidfilename = NULL;
	snprintf(config_path, XS_PATH_MAX, "%s/httpd.conf", calcpath(rootdir));

	if ((conffile = fopen(config_path, "r")))
	{
		while (fgets(buffer, BUFSIZ, conffile))
		{
			if (strncasecmp(buffer, "PidFile", 7))
				continue;
			for (p = buffer + 7; *p && isspace((int)*p); p++)
				/* DO NOTHING */;
			STRDUP(*pidfilename, p);
			for (p = *pidfilename; *p; p++)
				if (isspace((int)*p))
					*p = '\0';
			break;
		}
		fclose(conffile);
	}

	if (!*pidfilename)
		STRDUP(*pidfilename, calcpath(PID_PATH));
}

static	void
loadpidfile(const char *pidname)
{
	char	buffer[BUFSIZ];
	FILE	*pidfile;

	if ((pidfile = fopen(pidname, "r")))
	{
		if (!fgets(buffer, BUFSIZ, pidfile))
			errx(1, "PID line in `%s' is corrupt\n", pidname);
		else
		{
			httpdpid = (pid_t) atol(buffer);
			if (!fgets(startparams, BUFSIZ, pidfile))
				errx(1, "Arguments line in `%s' is corrupt\n",
				     pidname);
		}
		fclose(pidfile);
	}
	else
	{
		warn("fopen(`%s')", pidname);
		httpdpid = 0;
		strlcpy(startparams, BINDIR "/httpd", BUFSIZ);
	}
}

int
main(int argc, char **argv)
{
	char		buffer[BUFSIZ];
	char		*pidfilename = NULL;
	char		*cmd = NULL;
	int		option;

	strlcpy(rootdir, ROOT_DIR, BUFSIZ);
	while ((option = getopt(argc, argv, "d:p:v")) != EOF)
	{
		switch(option)
		{
		case 'd':
			strlcpy(rootdir, optarg, XS_PATH_MAX);
			break;
		case 'p':
			pidfilename = optarg;
			break;
		case 'v':
			cmd = "version";
			break;
		default:
			err(1, "Usage: %s [-d rootdir] [-p pidfile]", argv[0]);
		}
	}
	if (!pidfilename)
		getpidfilename(&pidfilename);

	if (argc != optind)
		cmd = argv[optind];

	if (cmd)
	{
		loadpidfile(pidfilename);
		control(cmd);
		return 0;
	}

	for (;;)
	{
		printf("httpdc> "); fflush(stdout);
		if (!fgets(buffer, BUFSIZ, stdin))
			break;
		while (buffer[0] && (buffer[strlen(buffer) - 1] <= ' '))
			buffer[strlen(buffer) - 1] = 0;
		if (!buffer[0])
			continue;
		loadpidfile(pidfilename);
		control(buffer);
	}
	fprintf(stderr, "End of input received\n");
	return 0;
}
